server/chat-plugins/chatlog.ts

```diff
@@ -166,10 +165,69 @@ export const LogReader = new class {
 		const prevMonth = new Date(new Date(`${month}-15`).getTime() - 30 * DAY);
 		return prevMonth.toISOString().slice(0, 7);
 	}
-
 	today() {
 		return Chat.toTimestamp(new Date()).slice(0, 10);
 	}
+	compareDates(a: string, b: string) {
+		const aDate = new Date(a).getTime();
+		const bDate = new Date(b).getTime();
+		return bDate - aDate;
+	}
+	isDate(item: string) {
+		return !isNaN(new Date(item).getTime());
+	}
+	async getBattleLog(tier: ID, number: number) {
+		// binary search!
+		const months = (await FS('logs').readdir())
+			.filter(f => this.isDate(f) && FS(`logs/${f}/${tier}`).existsSync())
+			.sort(this.compareDates);
+		if (!months.length) return;
+
+		const getBattleNum = (battleName: string) => Number(battleName.split('-')[1].slice(0, -9));
+
+		const getMonthRange = async (month: string) => {
+			const files = (await FS(`logs/${month}/${tier}/`).readdir()).sort(this.compareDates);
+			const firstDay = files[0];
+			const firstPath = FS(`logs/${month}/${tier}/${firstDay}`);
+			const firstFiles = await firstPath.readdir();
+			if (files.length === 1) {
+				const num = getBattleNum(firstFiles[0]);
+				return [num, num];
+			}
+			const lastDay = files[files.length - 1];
+
+			const secondPath = FS(`logs/${month}/${tier}/${lastDay}`);
+			const firstResult = getBattleNum(firstFiles[0]);
+			const lastFiles = await secondPath.readdir();
+			const secondResult = getBattleNum(lastFiles[lastFiles.length - 1]);
```

Comment:
You did not sort `lastFiles`, also I think `lastDayFiles` and `lastBattleNum` would be better variable names here.

Reply:
Also this should probably not be one algorithm. What we need are:

1. the first day of the first month
2. the last day of the last month
3. battle number ranges for a single day

We never have use for battle number ranges for an entire month; you're doing more readdirs than necessary and readdirs of a folder containing 100K files are expensive.